builtin/credential/aws/path_role.go

```diff
@@ -320,6 +322,142 @@ func (b *backend) setRole(ctx context.Context, s logical.Storage, roleName strin
 	return nil
 }
 
+// initialize is used to initialize the AWS roles
+func (b *backend) initialize(ctx context.Context, s logical.Storage) error {
+
+	// Initialize only if we are either:
+	//   (1) A local mount.
+	//   (2) Are _not_ a replicated standby cluster.
+	if b.System().LocalMount() || !b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary|consts.ReplicationPerformanceStandby) {
+
+		logger := b.Logger().Named("initialize")
+
+		// grab the guard
+		if !atomic.CompareAndSwapUint32(b.initializeCASGuard, 0, 1) {
+			logger.Error("initialization already in progress")
+			return nil
+		}
+
+		// check if this method has already been called
+		if b.isInitialized {
+			atomic.StoreUint32(b.initializeCASGuard, 0)
+			logger.Error("already initialized")
+			return nil
+		}
+		b.isInitialized = true
+
+		// kick off the role upgrader
+		go func() {
+			defer atomic.StoreUint32(b.initializeCASGuard, 0)
+
+			// Don't cancel when the original client request goes away
+			ctx = context.Background()
+
+			logger.Info("starting initialization")
+
+			updated, err := b.updateUpgradableRoleEntries(ctx, s)
+			if err != nil {
+				logger.Error("error running initialization", "error", err)
+				return
+			}
+			if !updated {
+				logger.Info("upgrade has already been performed for role storage version",
+					currentRoleStorageVersion)
+			}
+
+			logger.Info("initialization succeeded")
+		}()
+
+		return nil
+	}
+
+	return nil
+
+}
+
+// updateUpgradableRoleEntries upgrades and persists all of the role entries
+// that are in need of being upgraded.
+func (b *backend) updateUpgradableRoleEntries(ctx context.Context, s logical.Storage) (bool, error) {
+
+	// check if we've already upgraded to the current role storage
+	// version
+	version, err := b.upgradedRoleStorageVersion(ctx, s)
+	if err != nil {
+		return false, err
+	}
+	if version == currentRoleStorageVersion {
+		return false, nil
+	}
+
+	// Read all the role names.  We don't need to grab the mutex here for
+	// listing. The reason is that if we're in the process of creating a new
+	// mount, it will already happen on the active node (since it's a write)
+	// and will properly use the latest role structure.
+	roleNames, err := s.List(ctx, "role/")
+	if err != nil {
+		return false, err
+	}
+
+	// Upgrade the roles as necessary.
+	for _, roleName := range roleNames {
+		err := b.updateUpgradableRoleEntry(ctx, s, roleName)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// save the current role storage version
+	err = b.setUpgradedRoleStorageVersion(ctx, s, currentRoleStorageVersion)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// updateUpgradableRoleEntry uses the write lock to call roleInternal(), which
```

Comment:
There's no reason to keep grabbing the lock and giving it up, since this is all running synchronously before client requests. Just grab the lock once in the upgrade function above and call roleInternal directly.